File: .claude/skills/pr-sync/scripts/store.py
"""Store: the persistence seam. Skills go through this; they never
`open("state/...")` directly.

Two namespaces live under the state dir, and the split is load-bearing:

  cache/   100% reconstructable from GitHub (PRRecords, the relevance profile,
           the PR-detail SQLite, the workup SOAPs). Losing it costs recompute,
           never correctness — so it is best-effort and prunable.
  ledger/  the drafts and the accumulated draft-vs-posted learning series (§5).
           This exists *nowhere else* — GitHub never saw the drafts, by design —
           so it is durable: written once per cycle, never pruned.

Today the backend is the local filesystem (FileStore); a remote routine wraps
the same `state/` tree with a git round-trip (see state_sync.py) so it survives a
fresh clone. The seam is what lets the backend become LanceDB-on-object-storage
later without touching a single caller.

Keys are POSIX-style and MUST start with `cache/` or `ledger/`. That prefix is
how the backend decides what to prune vs keep.
"""

from __future__ import annotations

import json
import os
from typing import Any

CACHE = "cache"
LEDGER = "ledger"


def default_state_dir() -> str:
    here = os.path.dirname(os.path.abspath(__file__))
    repo_root = os.path.abspath(os.path.join(here, "..", "..", "..", ".."))
    return os.path.join(repo_root, "state")


def _validate_key(key: str) -> None:
    if not key or key.startswith("/") or "\\" in key:
        raise ValueError(f"bad store key: {key!r}")
    head = key.split("/", 1)[0]
    if head not in (CACHE, LEDGER):
        raise ValueError(f"store key must start with cache/ or ledger/: {key!r}")
    if ".." in key.split("/"):
        raise ValueError(f"store key may not contain '..': {key!r}")


class FileStore:
    """Filesystem-backed store rooted at a state dir. The whole tree is what the
    git backend round-trips, so anything written here survives a remote run iff
    state_sync persists it."""

    def __init__(self, root: str | None = None):
        self.root = root or default_state_dir()

    def local_path(self, key: str, *, make_parents: bool = False) -> str:
        """Materialize a real filesystem path for `key`. Needed by callers that
        can't take bytes (e.g. sqlite3.connect)."""
        _validate_key(key)
        path = os.path.join(self.root, *key.split("/"))
        if make_parents:
            os.makedirs(os.path.dirname(path), exist_ok=True)
        return path

    def exists(self, key: str) -> bool:
        return os.path.exists(self.local_path(key))

    def get_json(self, key: str) -> Any | None:
        path = self.local_path(key)
        if not os.path.exists(path):
            return None
        try:
            with open(path, encoding="utf-8") as fh:
                return json.load(fh)
        except (OSError, ValueError):
            return None

    def put_json(self, key: str, obj: Any) -> None:
        path = self.local_path(key, make_parents=True)
        with open(path, "w", encoding="utf-8") as fh:
            json.dump(obj, fh, indent=2)

    def get_text(self, key: str) -> str | None:
        path = self.local_path(key)
        if not os.path.exists(path):
            return None
        try:
            with open(path, encoding="utf-8") as fh:
                return fh.read()
        except OSError:
            return None

    def put_text(self, key: str, text: str) -> None:
        path = self.local_path(key, make_parents=True)
        with open(path, "w", encoding="utf-8") as fh:
            fh.write(text)

    def delete(self, key: str) -> bool:
        path = self.local_path(key)
        if os.path.exists(path):
            os.remove(path)
            return True
        return False

    def list_keys(self, prefix: str) -> list[str]:
        """All file keys under `prefix` (a key prefix, e.g. 'ledger/dispositions'
        or 'cache/workups'), POSIX-style, sorted oldest-name first."""
        _validate_key(prefix.rstrip("/") + "/x")  # prefix must be in a namespace
        base = os.path.join(self.root, *prefix.split("/"))
        if not os.path.isdir(base):
            return []
        out: list[str] = []
        for dirpath, _dirs, files in os.walk(base):
            rel_dir = os.path.relpath(dirpath, self.root)
            for fn in files:
                rel = os.path.join(rel_dir, fn)
                out.append(rel.replace(os.sep, "/"))
        return sorted(out)


# --- artifact key registry -------------------------------------------------
# One place that knows the on-disk layout, so the cache/ledger split and the
# swap to a different backend stay centralized.

RECORDS = "cache/records.json"
PANEL = "cache/panel.json"
RELEVANCE_PROFILE = "cache/relevance_profile.json"
PR_DETAIL_DB = "cache/pr_detail.sqlite"

DISPOSITIONS_PREFIX = "ledger/dispositions"
RECONCILE_AGREEMENT = "ledger/reconcile/agreement.json"


def _safe_segment(s: str) -> str:
    """Make a string safe as a single path segment (slashes, dots collapsed)."""
    return "".join(c if (c.isalnum() or c in "-_.") else "_" for c in s)


def disposition_key(cycle: str) -> str:
    return f"{DISPOSITIONS_PREFIX}/{_safe_segment(cycle)}.json"


def workup_key(repo: str, number: int, sha: str) -> str:
    """The workup cache key: a SOAP is valid only for the exact head SHA it was
    written against (§3 — a force-push invalidates it)."""
    owner, name = (repo.split("/", 1) + [""])[:2]
    return f"cache/workups/{_safe_segment(owner)}/{_safe_segment(name)}/{int(number)}/{_safe_segment(sha)}.json"


def embedding_key(model: str, content_hash: str) -> str:
    """Cache key for an embedding vector. Scoped by model name so swapping the
    embedding model never serves stale vectors from a different one."""
    return f"cache/embeddings/{_safe_segment(model)}/{_safe_segment(content_hash)}.json"

File: internal/store/store.go
// Package store is the persistence seam. Callers go through it; they never open
// state files directly. Two namespaces live under the state dir and the split
// is load-bearing:
//
//	cache/   100% reconstructable from GitHub (records, relevance profile, the
//	         PR-detail SQLite, workup SOAPs). Best-effort and prunable.
//	ledger/  drafts + the accumulated learning series. Exists nowhere else, so
//	         it is durable: written once per cycle, never pruned.
//
// Keys are POSIX-style and MUST start with cache/ or ledger/. That prefix is how
// the backend decides what to prune vs keep, and the seam is what lets the
// backend swap to another store later without touching a caller.
package store

import (
	"encoding/json"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

const (
	nsCache  = "cache"
	nsLedger = "ledger"
)

// Artifact key registry: the one place that knows the on-disk layout.
const (
	RecordsKey          = "cache/records.json"
	PanelKey            = "cache/panel.json"
	RelevanceProfileKey = "cache/relevance_profile.json"
	PRDetailDBKey       = "cache/pr_detail.sqlite"

	DispositionsPrefix    = "ledger/dispositions"
	ReconcileAgreementKey = "ledger/reconcile/agreement.json"
)

func validateKey(key string) error {
	if key == "" || strings.HasPrefix(key, "/") || strings.Contains(key, `\`) {
		return fmt.Errorf("bad store key: %q", key)
	}
	head := key
	if i := strings.IndexByte(key, '/'); i >= 0 {
		head = key[:i]
	}
	if head != nsCache && head != nsLedger {
		return fmt.Errorf("store key must start with cache/ or ledger/: %q", key)
	}
	for _, seg := range strings.Split(key, "/") {
		if seg == ".." {
			return fmt.Errorf("store key may not contain '..': %q", key)
		}
	}
	return nil
}

// FileStore is a filesystem-backed store rooted at a state dir.
type FileStore struct {
	Root string
}

// New returns a FileStore rooted at root.
func New(root string) *FileStore {
	return &FileStore{Root: root}
}

// LocalPath materializes a real filesystem path for key. Needed by callers that
// can't take bytes (e.g. sqlite). With makeParents it creates the parent dir.
func (s *FileStore) LocalPath(key string, makeParents bool) (string, error) {
	if err := validateKey(key); err != nil {
		return "", err
	}
	path := filepath.Join(append([]string{s.Root}, strings.Split(key, "/")...)...)
	if makeParents {
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return "", err
		}
	}
	return path, nil
}

// Exists reports whether key is present (false on a bad key).
func (s *FileStore) Exists(key string) bool {
	path, err := s.LocalPath(key, false)
	if err != nil {
		return false
	}
	_, err = os.Stat(path)
	return err == nil
}

// GetJSON decodes key into v. found=false when the file is missing or corrupt
// (mirroring the Python producer, which tolerates a corrupt cache file).
func (s *FileStore) GetJSON(key string, v any) (found bool, err error) {
	path, err := s.LocalPath(key, false)
	if err != nil {
		return false, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, err
	}
	if err := json.Unmarshal(data, v); err != nil {
		return false, nil
	}
	return true, nil
}

// PutJSON writes v as indented JSON, creating parents.
func (s *FileStore) PutJSON(key string, v any) error {
	path, err := s.LocalPath(key, true)
	if err != nil {
		return err
	}
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

// GetText reads key as text; found=false when missing.
func (s *FileStore) GetText(key string) (text string, found bool, err error) {
	path, err := s.LocalPath(key, false)
	if err != nil {
		return "", false, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return "", false, nil
		}
		return "", false, err
	}
	return string(data), true, nil
}

// PutText writes text to key, creating parents.
func (s *FileStore) PutText(key, text string) error {
	path, err := s.LocalPath(key, true)
	if err != nil {
		return err
	}
	return os.WriteFile(path, []byte(text), 0o644)
}

// Delete removes key, reporting whether it existed.
func (s *FileStore) Delete(key string) (bool, error) {
	path, err := s.LocalPath(key, false)
	if err != nil {
		return false, err
	}
	if _, err := os.Stat(path); err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, err
	}
	if err := os.Remove(path); err != nil {
		return false, err
	}
	return true, nil
}

// ListKeys returns all file keys under prefix, POSIX-style, sorted.
func (s *FileStore) ListKeys(prefix string) ([]string, error) {
	if err := validateKey(strings.TrimRight(prefix, "/") + "/x"); err != nil {
		return nil, err
	}
	base := filepath.Join(append([]string{s.Root}, strings.Split(prefix, "/")...)...)
	info, err := os.Stat(base)
	if err != nil || !info.IsDir() {
		return []string{}, nil
	}
	var out []string
	err = filepath.WalkDir(base, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(s.Root, path)
		if err != nil {
			return err
		}
		out = append(out, filepath.ToSlash(rel))
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(out)
	if out == nil {
		out = []string{}
	}
	return out, nil
}

// safeSegment makes a string safe as a single path segment.
func safeSegment(s string) string {
	var b strings.Builder
	for _, c := range s {
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' {
			b.WriteRune(c)
		} else {
			b.WriteByte('_')
		}
	}
	return b.String()
}

// DispositionKey is the ledger key for a cycle's drafted dispositions.
func DispositionKey(cycle string) string {
	return DispositionsPrefix + "/" + safeSegment(cycle) + ".json"
}

// WorkupKey is the workup-cache key: a SOAP is valid only for the exact head
// SHA it was written against (a force-push invalidates it).
func WorkupKey(repo string, number int, sha string) string {
	owner, name := repo, ""
	if i := strings.IndexByte(repo, '/'); i >= 0 {
		owner, name = repo[:i], repo[i+1:]
	}
	return "cache/workups/" + safeSegment(owner) + "/" + safeSegment(name) + "/" +
		strconv.Itoa(number) + "/" + safeSegment(sha) + ".json"
}

// EmbeddingKey is the cache key for an embedding vector, scoped by model name.
func EmbeddingKey(model, contentHash string) string {
	return "cache/embeddings/" + safeSegment(model) + "/" + safeSegment(contentHash) + ".json"
}
